### DISCUSSION TITLE
Jepsen for JetStream

### DISCUSSION THREAD

**Discussion by jessekv** [nats-io/nats-server#3312] [General]
Earlier this year Redpanda went through the Jepsen audit process: http://jepsen.io/analyses/redpanda-21.10.1

Wondering if there is a plan to do something similar for JetStream?

**Comment by derekcollison** [comment]
Yes there is.

**Comment by jessekv** [comment]
That's great, I'm really looking forward to reading about it! Is there a rough time frame yet for when it will happen?

**Reply by derekcollison** [reply]
Possibly 2.9, or 3.0 release. 2.9 is scheduled for next week, fingers crossed.

The testing process is quite pricey as well, so need to budget.

**Comment by hwinkel** [comment]
Is there any progress in this regards?

**Reply by derekcollison** [reply]
Yes we have been working with a stealth startup for almost a year now that is incredible. Will let @mprimi divulge more.

**Reply by mprimi** [reply]
I'm not going to comment on the official Jepsen audit, but I can tell you that for some time we have been hammering NATS and JS with a Jepsen-like approach (e.g. invariant testing with fault injection). Jepsen(-like) testing is great to uncover bugs in code that deals with fault-tolerance and recovery, that are hard to test otherwise.
We've also been relying on Antithesis (https://antithesis.com/) find even more insidious issues. Antithesis is like Jepsen on steroids and can help to seek and destroy one-in-a-million heisenbugs that are so hard to track down in distributed systems.

**Comment by atropos112** [comment]
With it being couple months, have there been any news about audits?

**Reply by mprimi** [reply]
We keep continue expanding on testing. 

What I can tell you is we're putting NATS/JetStreams through scenarios and trials much more evil, complex and comprehensive than what Jepsen (the software) would do. 
Happy to share privately, but we're not ready to be completely open about this just yet.

On the other side, you probably want to see a third-party report and stamp of approval (i.e. Jepsen the blog).
I don't have comments or an ETA on that.

**Reply by bbigras** [reply]
It would be nice to have something soon, even if as basic as Jepsen.

I'm not sure how to interpret [this comment from the Jepsen dev](https://news.ycombinator.com/item?id=42117147) but it kinda sounds like jetstream has a bad reputation or something.

**Reply by hpvd** [reply]
@bbigras this link gives also the right context/question (its the parent of yours) https://news.ycombinator.com/item?id=42117108

**Reply by roded** [reply]
@bbigras @hpvd I'm still not sure what conclusion to draw from that discussion. I read it as @bbigras did.

**Comment by derekcollison** [comment]
We have invested heavily into [Antithesis](https://antithesis.com/) for this type of testing, since it is deterministic and Jepsen is not. We have been working with them as one of their first customers for over 2 years now and have made great progress.

If a Jepsen dev wants to run NATS / JetStream through Jepsen that would be great, by for our investment we will stick with Antithesis.

**Reply by roded** [reply]
Hi @derekcollison , any plans on making an Antithesis report public or the like?
Thanks

**Reply by derekcollison** [reply]
Antithesis is a tool, it does not run tests and generate reports on its own. We do have plans to shed more light on our own extensive testing with Antithesis.

**Reply by roded** [reply]
Thanks @derekcollison . Looking forward for any such publication.

**Comment by aphyr** [comment]
Just to follow up, I spent a couple months writing tests for NATS, and put together a small report here: https://jepsen.io/analyses/nats-2.12.1

**Reply by derekcollison** [reply]
Thanks Kyle. Much appreciated and the product and ecosystem will be better for it.

**Reply by aphyr** [reply]
Y'all are welcome! If you would like to look at membership changes, feel free to write--a lot of the skeleton for testing membership is there, and it can definitely break things, but @MauriceVanVeen and I are pretty sure it's being too aggressive in the process.

**Reply by MauriceVanVeen** [reply]
We've since fixed several issues relating to peer-removes that, although usable for a human, made it hard to automate safely. The most important one being that we now don't respond 'success, peer-removed' immediately, but instead wait for quorum: https://github.com/nats-io/nats-server/pull/7581. We also added protections to not allow concurrent membership changes and other edge cases: https://github.com/nats-io/nats-server/pull/7565, https://github.com/nats-io/nats-server/pull/7589, https://github.com/nats-io/nats-server/pull/7600, https://github.com/nats-io/nats-server/pull/7602, https://github.com/nats-io/nats-server/pull/7609, https://github.com/nats-io/nats-server/pull/7610. All of these changes will be included in 2.12.3

I'm fairly confident that makes doing peer-removes safer, the remaining bit will likely be the way how the nemesis keeps track of the membership view as well as retrying failed peer-remove requests while maintaining quorum, etc. (where currently it's a bit too aggressive and can break quorum by ending up with a too large cluster due to adding too many new nodes while the old ones weren't peer-removed successfully yet).
